microsoft/testsuites/dpdk/dpdkutil.py

```diff
@@ -610,3 +619,455 @@ def _parallel_cleanup(node: Node) -> None:
+        )
+
+    # we're about to fully ruin this environment, so mark these dirty before we start
+    for node in [forwarder, sender, receiver]:
+        node.mark_dirty()
+
+    # enable ip forwarding on secondary and tertiary nics
+    run_in_parallel(
+        [partial(__enable_ip_forwarding, node) for node in environment.nodes.list()]
+    )
+
+    # organize our nics by subnet.
+    # NOTE: we're ignoring the primary interfaces on each VM since we need it
+    #  to service the ssh connection.
+    # Subnet A is actually the secondary NIC.
+    #  Subnet B is actually the tertiary NIC.
+    # For the test, don't sweat it. A is send side, B is receive side.
+    subnet_a_nics = {
+        forwarder: forwarder.nics.get_nic_by_index(send_side),
+        sender: sender.nics.get_nic_by_index(send_side),
+        receiver: receiver.nics.get_nic_by_index(send_side),
+    }
+    subnet_b_nics = {
+        forwarder: forwarder.nics.get_nic_by_index(receive_side),
+        receiver: receiver.nics.get_nic_by_index(receive_side),
+        sender: sender.nics.get_nic_by_index(receive_side),
+    }
+
+    # We use ntttcp for snd/rcv which will respect the kernel route table.
+    # SO: remove the unused interfaces and routes which could skip the forwarder
+    unused_nics = {
+        sender: subnet_b_nics[sender],
+        receiver: subnet_a_nics[receiver],
+    }
+    reroute_traffic_and_disable_nic(
+        node=sender,
+        src_nic=unused_nics[sender],
+        dst_nic=subnet_b_nics[receiver],
+        new_gateway_nic=subnet_a_nics[forwarder],
+    )
+    reroute_traffic_and_disable_nic(
+        node=receiver,
+        src_nic=unused_nics[receiver],
+        dst_nic=subnet_a_nics[sender],
+        new_gateway_nic=subnet_b_nics[forwarder],
+    )
+
+    # AZ ROUTING TABLES
+    # The kernel routes are not sufficient, since Azure also manages
+    # the VNETs for the VMS. Azure doesn't really care about ethernet
+    # addresses and looks at IP, so we must set up an azure route table
+    # and apply it to our VNETs to send traffic to the DPDK forwarder.
+    # see:
+    # https://learn.microsoft.com/en-us/azure/virtual-network/
+    #  tutorial-create-route-table-portal#create-a-route
+    sender.features[NetworkInterface].create_route_table(
+        nic_name=subnet_a_nics[forwarder].name,
+        route_name="fwd-rx",
+        subnet_mask=ipv4_to_lpm(subnet_a_nics[sender].ip_addr),
+        em_first_hop=AZ_ROUTE_ALL_TRAFFIC,
+        next_hop_type="VirtualAppliance",
+        dest_hop=subnet_a_nics[forwarder].ip_addr,
+    )
+    receiver.features[NetworkInterface].create_route_table(
+        nic_name=subnet_b_nics[forwarder].name,
+        route_name="fwd-tx",
+        subnet_mask=ipv4_to_lpm(subnet_b_nics[receiver].ip_addr),
+        em_first_hop=AZ_ROUTE_ALL_TRAFFIC,
+        next_hop_type="VirtualAppliance",
+        dest_hop=subnet_b_nics[forwarder].ip_addr,
+    )
+
```

Comment:
can we use loop for above code?

Reply:
Theoretically, I don't get what the benefit would be. Each node is calling the function with arguments from the other nodes so we would just be removing the argument labels in the function call.